File: Utils/C_Tests/Bin_Tests.cpp
#include "../Catch2/single_include/catch2/catch.hpp"
#include "utils.hpp"
#include "wrappers.h"

using Catch::Matchers::Equals;

TEST_CASE("Given a file in an unrecognised format When it is parsed Then an exception is raised") {
	std::vector<uint8_t> inputFile = utils::ReadFile("badorder.lbm");
	bin_p bin = Test_Bin_create(inputFile.data(), uint32_t(inputFile.size()));
	REQUIRE(bin == NULL);
	REQUIRE(Test_Exception_code() == 2);
	REQUIRE_THAT(Test_Exception_string(), Equals("Unrecognised format"));
}

TEST_CASE("Given a file which is too short When it is parsed Then an exception is raised") {
	std::vector<uint8_t> inputFile = utils::ReadFile("tooshort.dat");
	bin_p bin = Test_Bin_create(inputFile.data(), uint32_t(inputFile.size()));
	REQUIRE(bin == NULL);
	REQUIRE(Test_Exception_code() == 3);
	REQUIRE_THAT(Test_Exception_string(), Equals("Buffer too small"));
}

TEST_CASE("Given a good file When it is converted to an unrecognised format Then an exception is raised") {
	std::vector<uint8_t> inputFile = utils::ReadFile("BASICU.bin");
	bin_p bin = Test_Bin_create(inputFile.data(), uint32_t(inputFile.size()));
	try {
		binBuffer_p uncompressedBuffer = Test_Bin_convert(bin, 7);
		REQUIRE(uncompressedBuffer == NULL);
		REQUIRE(Test_Exception_code() == 2);
		REQUIRE_THAT(Test_Exception_string(), Equals("Unrecognised format"));
	}
	catch (...) {
	}
	Test_Bin_destroy(bin);
}

TEST_CASE("Given a good file When an overlong comment is added Then an exception is raised") {
	std::vector<uint8_t> inputFile = utils::ReadFile("BASICU.bin");
	bin_p bin = Test_Bin_create(inputFile.data(), uint32_t(inputFile.size()));
	std::string comment(3001, 'x');
	uint8_t result = Test_Bin_setComment(bin, comment.c_str());
	REQUIRE(result == 1);
	REQUIRE(Test_Exception_code() == 3);
	REQUIRE_THAT(Test_Exception_string(), Equals("Comment is too large to be stored"));
	Test_Bin_destroy(bin);
}

TEST_CASE("Given a bin When the bin is constructed Then it is correctly initialised") {
	bin_p bin = Test_Bin_createNew();
	REQUIRE(bin->mapWidth == 20);
	REQUIRE(bin->mapHeight == 13);
	Test_Bin_destroy(bin);
}

TEST_CASE("Given a bin When the bin is converted to uncompressed Then it is correctly formed") {
	std::vector<uint8_t> inputFile = utils::ReadFile("BASIC.BIN");
	bin_p bin = Test_Bin_create(inputFile.data(), uint32_t(inputFile.size()));
	binBuffer_p uncompressedBuffer = NULL;
	try {
		uncompressedBuffer = Test_Bin_convert(bin, CONVERSION_UNCOMPRESSED);
		uint32_t bufferSize = Test_BinBuffer_size(uncompressedBuffer);
		std::vector<uint8_t> actualBuffer(bufferSize);
		uint8_t result = Test_BinBuffer_buffer(uncompressedBuffer, actualBuffer.data(), bufferSize);
		
		REQUIRE(bin->mapWidth == 25);
		REQUIRE(bin->mapHeight == 20);
		REQUIRE(bufferSize == 25526);
		REQUIRE(result == (uint8_t)true);
		REQUIRE(actualBuffer.size() == bufferSize);
		REQUIRE(actualBuffer[0] == 'b');
		REQUIRE(actualBuffer[1] == 'y');
		REQUIRE(actualBuffer[2] == 't');
		REQUIRE(actualBuffer[3] == '4');
	}
	catch (...) {

	}
	Test_Bin_destroy(bin);
	Test_BinBuffer_destroy(uncompressedBuffer);
}

TEST_CASE("Given a bin When the bin is converted to compressed Then it is correctly formed") {
	std::vector<uint8_t> inputFile = utils::ReadFile("BASIC.BIN");
	bin_p bin = Test_Bin_create(inputFile.data(), uint32_t(inputFile.size()));
	binBuffer_p uncompressedBuffer = NULL;
	try {
		uncompressedBuffer = Test_Bin_convert(bin, CONVERSION_COMPRESSED);
		uint32_t bufferSize = Test_BinBuffer_size(uncompressedBuffer);
		std::vector<uint8_t> actualBuffer(bufferSize);
		uint8_t result = Test_BinBuffer_buffer(uncompressedBuffer, actualBuffer.data(), bufferSize);
		
		REQUIRE(bin->mapWidth == 25);
		REQUIRE(bin->mapHeight == 20);
		REQUIRE(bufferSize < 25526);
		REQUIRE(result == (uint8_t)true);
		REQUIRE(actualBuffer.size() == bufferSize);
		REQUIRE(actualBuffer[0] == 'b');
		REQUIRE(actualBuffer[1] == 'y');
		REQUIRE(actualBuffer[2] == 't');
		REQUIRE(actualBuffer[3] == '4');
		REQUIRE(actualBuffer[4] < 10);
	}
	catch (...) {

	}
	Test_Bin_destroy(bin);
	Test_BinBuffer_destroy(uncompressedBuffer);
}

TEST_CASE("Given a yaml When the yaml is converted Then a comment can be extracted") {
	std::vector<uint8_t> inputFile = utils::ReadFile("BASIC.yml");
	bin_p bin = Test_Bin_create(inputFile.data(), uint32_t(inputFile.size()));
	try {
		char comment[3001];
		uint32_t result = Test_Bin_getComment(bin, comment, 3001);
		REQUIRE(result > 0);
		REQUIRE_THAT(comment, Equals("Basic YAML description of Fury of the Furries BIN file. Test asset.\nThis file should round-trip"));
	}
	catch (...) {
	}
	Test_Bin_destroy(bin);
}

TEST_CASE("Given a bin When a comment is added Then the comment can be extracted") {
	bin_p bin = Test_Bin_createNew();
	uint8_t result = Test_Bin_setComment(bin, "This is a test");
	REQUIRE(result == 0);
	try {
		char comment[100];
		uint32_t result = Test_Bin_getComment(bin, comment, 100);
		REQUIRE(result > 0);
		REQUIRE_THAT(comment, Equals("This is a test"));
	}
	catch (...) {
	}
	Test_Bin_destroy(bin);
}

File: Tests.Utils/Bin_Tests.cs
using carbon14.FuryStudio.Utils;
using System.Runtime.InteropServices;
using Xunit.Sdk;

namespace carbon14.FuryStudio.Tests.Utils
{
    public class Bin_Tests
    {
        [Fact]
        public void Given_a_valid_bin_file_When_a_BIN_object_is_created_Then_the_map_size_is_correct()
        {
            using (Bin bf = new(TestHelpers.ReadFile("BASIC.BIN"))) {
                Assert.Equal(25518, Marshal.SizeOf(bf.data));
                Assert.Equal(25, bf.data.mapWidth);
                Assert.Equal(20, bf.data.mapHeight);
            }

        }

        [Fact]
        public void Given_an_empty_bin_When_created_Then_the_map_size_is_correct()
        {
            using (Bin bf = new())
            {
                Assert.Equal(25518, Marshal.SizeOf(bf.data));
                Assert.Equal(20, bf.data.mapWidth);
                Assert.Equal(13, bf.data.mapHeight);
            }
        }

        [Fact]
        public void Given_a_valid_bin_When_it_is_converted_to_uncompressed_Then_the_buffer_is_correct()
        {
            using (Bin bf = new(TestHelpers.ReadFile("BASIC.BIN")))
            {
                byte[]? uncompressed = bf.Convert(Bin.ConversionType.Uncompressed);
                Assert.NotNull(uncompressed);
                Assert.Equal(25526, uncompressed.Length);
                Assert.Equal(25, bf.data.mapWidth);
                Assert.Equal(20, bf.data.mapHeight);
                using (Bin bf2 = new(uncompressed))
                {
                    byte[]? uc2 = bf2.Convert(Bin.ConversionType.Uncompressed);
                    Assert.Equal(uncompressed, uc2);
                }
            }
        }

        [Fact]
        public void Given_a_valid_bin_When_it_is_converted_to_compressed_Then_the_buffer_is_correct()
        {
            using (Bin bf = new(TestHelpers.ReadFile("BASIC.BIN")))
            {
                byte[]? compressed = bf.Convert(Bin.ConversionType.Compressed);
                Assert.NotNull(compressed);
                Assert.True(compressed.Length < 25526);
                Assert.Equal(25, bf.data.mapWidth);
                Assert.Equal(20, bf.data.mapHeight);
                using (Bin bf2 = new(compressed))
                {
                    byte[]? c2 = bf2.Convert(Bin.ConversionType.Compressed);
                    Assert.Equal(compressed, c2);
                }
            }
        }

        [Fact]
        public void Given_a_valid_yaml_When_a_comment_is_added_Then_the_comment_can_be_extracted()
        {
            using (Bin bf = new(TestHelpers.ReadFile("BASIC.yml")))
            {
                Assert.Equal("Basic YAML description of Fury of the Furries BIN file. Test asset.\nThis file should round-trip", bf.Comment);
            }
        }

        [Fact]
        public void Given_a_constructed_bin_When_a_comment_is_added_Then_the_comment_can_be_extracted()
        {
            using (Bin bf = new())
            {
                bf.Comment = "Test comment";
                Assert.Equal("Test comment", bf.Comment);
            }

        }

        [Fact]
        public void Given_an_invalid_bin_file_When_a_bin_is_created_Then_an_exception_is_raised()
        {
            try
            {
                throw Assert.Throws<FuryException>(() => { Bin bin = new(TestHelpers.ReadFile("badorder.lbm")); });
            }
            catch (FuryException ex)
            {
                Assert.Equal(ErrorCodes.UNSUPPORTED_FORMAT, ex.ErrorCode);
                Assert.Equal("Unrecognised format", ex.Message);
            }
        }

        [Fact]
        public void Given_a_valid_bin_file_When_it_is_converted_Then_it_will_correctly_round_trip()
        {
            byte[]? original = TestHelpers.ReadFile("BASIC.yml");
            byte[]? expected = TestHelpers.ReadFile("BASIC.BIN");
            byte[]? actual;
            using (Bin bin = new(original))
            {
                actual = bin.Convert(Bin.ConversionType.Compressed);
            }
            Assert.True(actual?.SequenceEqual(expected), "Buffer differs");
            Assert.True(actual?.SequenceEqual(expected), "Buffer differs");
        }

        [Fact]
        public void Given_a_valid_bin_When_it_is_serialised_and_deserialised_Then_it_matches()
        {
            byte[]? file;
            using (Bin bin = new())
            {
                bin.data.mapWidth = 25;
                bin.data.mapHeight = 20;
                bin.Comment = "This is a test comment";
                file = bin.Convert(Bin.ConversionType.Compressed);
            }
            Assert.NotNull(file);
            using (Bin bin = new(file))
            {
                Assert.Equal("This is a test comment", bin.Comment);
                Assert.Equal(25, bin.data.mapWidth);
                Assert.Equal(20, bin.data.mapHeight);
            }
        }
    }
}
